src/main/java/com/iota/iri/utils/collections/impl/BoundedSetWrapper.java

```diff
@@ -0,0 +1,173 @@
+
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+
+/**
+ * A set that doesn't allow to add elements to it once it is full
+ *
+ * @param <E> the type parameter
+ */
+public class BoundedSetWrapper<E> implements BoundedSet<E>{
+    private final int maxSize;
+    private final Set<E> delegate;
+
+    /**
+     * Wraps the given set {@code c}
+     *
+     * @param c the set which you delegate the actions to
+     * @param maxSize the max size
+     * @throws NullPointerException if the {@code c} is null
+     * @throws IllegalArgumentException if {@code c} is larger than {@code maxSize}
+     */
+    public BoundedSetWrapper(Set<E> c, int maxSize) {
+        Objects.requireNonNull(c, "trying to wrap a null set");
+        if (c.size() > maxSize) {
+            throw new IllegalArgumentException(String.format("The given set size %d is larger then maxSize %d", c.size(),
+                    maxSize));
+        }
+        this.maxSize = maxSize;
+        this.delegate = c;
+    }
+
+    @Override
+    public int getMaxSize() {
+        return maxSize;
+    }
+
+    @Override
+    public int size() {
+       return delegate.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return delegate.isEmpty();
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        return delegate.contains(o);
+    }
+
+    @Override
+    public Iterator<E> iterator() {
+        return delegate.iterator();
+    }
+
+    @Override
+    public void forEach(Consumer<? super E> action) {
+        delegate.forEach(action);
+    }
+
+    @Override
+    public Object[] toArray() {
+        return delegate.toArray();
+    }
+
+    @Override
+    public <T> T[] toArray(T[] a) {
+        return delegate.toArray(a);
+    }
+
+    @Override
+    public boolean add(E e) {
+        if (isFull()) {
+            Iterator<E> iterator = delegate.iterator();
+            iterator.next();
+            iterator.remove();
+        }
+
+        return delegate.add(e);
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        return delegate.remove(o);
+    }
+
+    @Override
+    public boolean containsAll(Collection<?> c) {
+        return delegate.containsAll(c);
+    }
+
+    @Override
+    public boolean addAll(Collection<? extends E> c) {
+        if (canCollectionBeFullyAdded(c)) {
+            return delegate.addAll(c);
+        }
+        else {
+            Set<E> transform = new HashSet<>(c);
+            Set<E> difference = SetUtils.difference(delegate, transform).toSet();
+            if (difference.size() > getMaxSize()) {
+                throw new IllegalArgumentException(String.format("The given set size %d is larger then maxSize %d",
+                        difference.size(), maxSize));
+            }
+            int itemsToDelete = delegate.size() + difference.size() - getMaxSize();
```

Comment:
this violates the description `A set that doesn't allow to add elements to it once it is full` as you're 
deleting items to make room

Reply:
ditto